README.md
<img src="https://i.imgur.com/MiGH9tW.png" alt="Flutterbook" />

# Flutterbook

A storyboarding tool to accelerate the development of Flutter widgets. Inspired by Storybook.js.

**The package is built to support Windows and Web layouts. There is no support for mobile support at the moment. This is new and still under heavy development, expect many changes to occur with usage.**

<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-purple.svg" alt="License: MIT"></a>

---

Go subscribe to my YouTube channel to support development.

# [![MOONSDONTBURN Header](https://i.imgur.com/1QHjcUZ.png)](https://www.youtube.com/channel/UCurQRmT17EyOIrdPseiastg)

### Demo

<a href="https://flutterbook.vercel.app/#/" target="_blank">Demo Link Here</a>

## Overview ✨

#### Creating the Main

Start by creating the storyboard widget. This will wrap the Flutterbook widget
so that when it is edited, it will hot reload.

```dart
class Storyboard extends StatelessWidget {
  const Storyboard({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FlutterBook(
      categories: [
          // organizers ...
      ],
    );
  }
}
```

After creating the Storyboard widget, you can run it by using the main. I recommend making another
file named `main_storyboard.dart` to maintain your storyboard.

```dart
void main() {
  runApp(const Storyboard());
}
// Or run it directly, but this will cost the hot reload.
// void main() {
//   runApp(const Flutterbook(categories:[...]));
// }
```

#### How to add Widgets

To add widgets, you simply put `Organizers`, aka `Category`, `Folder`, `Component`, and `ComponentState`.

The hierarchy goes

`Category` > `Folder` > `Component` > `ComponentState`

but you may also do

`Category` > `Component` > `ComponentState`

or

`Category` > `Folder` > `Folder` > `Component` > `ComponentState`

Here is a complex example of a single `Category`. In the example, we use the builder
where we are able to access different controls.

```dart
Category(
    categoryName: 'LIBRARY',
    organizers: [
    Folder(
        folderName: 'Charts',
        organizers: [
        Component(
            componentName: 'LineGraph',
            states: [
            ComponentState.child(
                stateName: 'Default Container',
                child: Container(),
            ),
            ],
        ),
        ],
    ),
    Component(
        componentName: 'Button',
        compenentMarkdoown: """##A global button component""",
        states: [
        ComponentState(
            stateName: 'Primary',
            markdown: """Used for the main action to be performed"""
            codeSample:  r'''
Button.primary(
  child: Text('Primary Button'),
  onPressed: () {},
);
'''
            builder: (context, c) {
            return Center(
                child: SizedBox(
                width: c.number(
                    label: 'Number',
                    initial: 50,
                    min: 50,
                    max: 250,
                ),
                height: c.number(
                    label: 'height',
                    initial: 50,
                    min: 50,
                    max: 250,
                    description: 'random stuff',
                ),
                child: CupertinoButton.filled(
                    onPressed: c.boolean(
                    label: 'boolean',
                    initial: true,
                    )
                        ? () {}
                        : null,
                    child: Text(
                    c.text(label: 'Text', initial: 'Hello World'),
                    ),
                ),
                ),
            );
            },
        ),
        ],
    ),
    ],
),
```

#### That's it!

## Roadmap 🚧

- [ ] Mobile Support
- [ ] 100% Code Coverage Testing
- [X ] Documentation Support in ComponentState's
- [ ] Better Docs
- [X ] Device Previews
- [ ] Shareable Handoffs
- [ ] Optimization

## Contributors 🔥

_Your name could be here_ 😉

<a href="https://github.com/GhostWalker562/flutterbook/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=GhostWalker562/flutterbook" />
</a>


CHANGELOG.md
## 0.1.0

- Initial release.

## 0.1.3

\* Header/branding is now swappable with another widget of choice

\* Scaling a widget will now clip according to the Canvas

\* Refactor to fit new naming scheme

- More components have been documented

## 0.1.4

- Fix navigation path from excluding the direct parent of the component state

# 0.1.5

- Ignore case during search

- Dependency Updates

# 0.1.6

- Device Preview
- Add Document Preview Page
- Add Markdown capabilites for `ComponentState`
- Fixed bug with Controls not working

# 0.1.7

- Add DartCodeView for code snippets
- Add padding to the Docs tab
- Add support for Multi Theming (3+ themes)
- Add toggle for panning
- Add Markdown capabilites for `Component`
- Fixed bug in example project component
- _BREAKING_ Renamed `DarkThemeProvider` to `ThemeProvider`

# 0.1.8

- Add Controls for Items

# 0.1.9

- Update dependencies
- Export l10n-related properties


example/lib/main.dart
import 'package:example/list_example.dart';
import 'package:example/themes.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutterbook/flutterbook.dart';

void main() {
  runApp(const Storyboard());
}

class Storyboard extends StatelessWidget {
  const Storyboard({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FlutterBook(
      themes: Themes.themes,
      categories: [
        Category(
          categoryName: 'LIBRARY',

          organizers: [
            Folder(
              folderName: 'Charts',
              organizers: [
                Component(
                  componentName: 'LineGraph',
                  states: [
                    ComponentState.child(
                      stateName: 'Default Container',
                      child: Container(),
                    ),
                  ],
                ),
              ],
            ),
            Component(
              componentName: 'Button',
              componentMarkdown: """
## Component Markdown Example
The `Button` is global a component used for user actions.
               """,
              states: [
                ComponentState(
                  markdown: """
### Component State Markdown Example
The `Button.primary` is used for the main action to be performed.
               """,
                  codeSample: r'''
Button.primary(
  child: Text('Primary Button'),
  onPressed: () {},
);
''',
                  stateName: 'Primary',
                  builder: (context, c) {
                    return Center(
                      child: SizedBox(
                        width: c.number(
                          label: 'Number',
                          initial: 100,
                          min: 100,
                          max: 250,
                        ),
                        height: c.number(
                          label: 'height',
                          initial: 50,
                          min: 50,
                          max: 250,
                          description: 'random stuff',
                        ),
                        child: CupertinoButton(
                          color: c.list<Color>(
                            label: "Color",
                            initial: Colors.red,
                            value: Colors.red,
                            list: [
                              ListItem(title: "Red", value: Colors.red),
                              ListItem(title: "Blue", value: Colors.blue),
                              ListItem(title: "Black", value: Colors.black),
                              ListItem(title: "Amber", value: Colors.amber),
                            ],
                          ),
                          onPressed: c.boolean(
                            label: 'boolean',
                            initial: true,
                          )
                              ? () {}
                              : null,
                          child: Text(
                            c.text(label: 'Text', initial: 'Hello World'),
                          ),
                        ),
                      ),
                    );
                  },
                ),
                ComponentState(
                  markdown: """
### Component State Markdown Example
The `Button.secondary` is used for user alternate actions.
               """,
                  codeSample: r'''
Button.secondary(
  child: Text('Primary Button'),
  onPressed: () {},
);
''',
                  stateName: 'Secondary',
                  builder: (context, c) {
                    return Center(
                      child: SizedBox(
                        width: c.number(
                          label: 'Number',
                          initial: 100,
                          min: 100,
                          max: 250,
                        ),
                        height: c.number(
                          label: 'height',
                          initial: 50,
                          min: 50,
                          max: 250,
                          description: 'random stuff',
                        ),
                        child: CupertinoButton(
                          onPressed: c.boolean(
                            label: 'boolean',
                            initial: true,
                          )
                              ? () {}
                              : null,
                          child: Text(
                            c.text(
                              label: 'Text',
                              initial: 'Hello World',
                            ),
                          ),
                        ),
                      ),
                    );
                  },
                ),
              ],
            ),
            Component(
              componentName: 'List',
              states: [
                ComponentState(
                  codeSample: """ListExample()""",
                  stateName: 'Primary',
                  builder: (context, c) => ListExample(),
                ),
              ],
            ),
          ],

        ),
      ],
    );
  }
}
